File: 1/2020.09.28/226.Invert Binary Tree/python/solution.py
class Solution:
    def invertTree(self, root: TreeNode) -> TreeNode:
        # DFS, recursively
        if not root:
            return None
        left = root.left
        root.left = self.invertTree(root.right)
        root.right = self.invertTree(left)
        return root

    def invertTree(self, root: TreeNode) -> TreeNode:
        # DFS, iteratively
        if not root:
            return None
        stack = [root]
        while stack != []:
            node = stack.pop()
            node.left, node.right = node.right, node.left
            if node.left:
                stack.append(node.left)
            if node.right:
                stack.append(node.right)
        return root

    def invertTree(self, root: TreeNode) -> TreeNode:
        # BFS
        if not root:
            return None
        q = [root]
        while q != []:
            node = q.pop(0)
            node.left, node.right = node.right, node.left
            if node.left:
                q.append(node.left)
            if node.right:
                q.append(node.right)
        return root

File: 1/2020.09.28/226.Invert Binary Tree/java/solution.java
class Solution {
    public TreeNode invertTree(TreeNode root) {
        // DFS
        if (root == null) {
            return null;
        }
        TreeNode left = root.left;
        root.left = invertTree(root.right);
        root.right = invertTree(left);
        return root;
    }

    public TreeNode invertTree(TreeNode root) {
        // iteratively, stack
        if (root == null) {
            return null;
        }
        Stack<TreeNode> stack = new Stack<>();
        stack.push(root);
        while (!stack.isEmpty()) {
            TreeNode node = stack.pop();
            TreeNode left = node.left;
            node.left = node.right;
            node.right = left;
            if (node.left != null) {
                stack.push(node.left);
            }
            if (node.right != null) {
                stack.push(node.right);
            }
        }
        return root;
    }

    public TreeNode invertTree(TreeNode root) {
        // BFS
        if (root == null) {
            return null;
        }
        Deque<TreeNode> q = new LinkedList<>();
        q.offerLast(root);
        while (!q.isEmpty()) {
            TreeNode node = q.pollFirst();
            TreeNode left = node.left;
            node.left = node.right;
            node.right = left;
            if (node.left != null) {
                q.offerLast(node.left);
            }
            if (node.right != null) {
                q.offerLast(node.right);
            }
        }
        return root;
    }
}
